File: leetcode-go/Q001-Q100/Q039-Combination-Sum/Q039-Combination-Sum.go
package main //Q039_Combination_Sum
import (
	//	"fmt"
	//	"sort"
	//	"reflect"
	"fmt"
	"sort"
)

func combinationSum(candidates []int, target int) [][]int {
	ans := [][]int{}
	sort.Ints(candidates)
	dfs(&ans, candidates, []int{}, target, 0)
	return ans
}

func dfs(ans *[][]int, candidates, combs []int, target int, begin int) {
	if target == 0 {
		*ans = append(*ans, combs)
		return
	}
	for i := begin; i < len(candidates) && target >= candidates[i]; i++ {
		dfs(ans, candidates, append(combs, candidates[i]), target-candidates[i], i)
	}
}

func main() {
	//combinationSum([]int{33,28,25,45,26,27,47,29,32,21,37,35,48,49,40,39,41,46,20,24,30,36,38,44,23,34}, 51)
	fmt.Println(combinationSum([]int{1, 2}, 4))
}

File: leetcode-java/src/Q001_Q099/Q039_Combination_Sum.java
package Q001_Q099;

import java.util.Arrays;
import java.util.LinkedList;
import java.util.List;

/**
 * Created by mqliang on 6/4/17.
 */
public class Q039_Combination_Sum {
    public List<List<Integer>> combinationSum(int[] candidates, int target) {
        Arrays.sort(candidates);
        List<List<Integer>> ans = new LinkedList<>();
        dfs(ans, new LinkedList<Integer>(), candidates, target, 0);
        return ans;
    }

    public void dfs(List<List<Integer>> ans, List<Integer> comb,  int[] candidates, int target, int begin) {
        if (target == 0) {
            ans.add(new LinkedList(comb));
            return;
        }
        for (int i = begin; i < candidates.length && target >= candidates[i]; i++) {
            comb.add(candidates[i]);
            dfs(ans, comb, candidates, target-candidates[i], i);
            comb.remove(comb.size()-1);
        }
    }
}
